src/coreclr/jit/morph.cpp

```diff
@@ -9273,6 +9296,127 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
+                                            &isRelative);
+
+    // Dereference the this pointer to obtain the method table, it is called vtab below
+    GenTree* vtab;
+    assert(VPTR_OFFS == 0); // We have to add this value to the thisPtr to get the methodTable
+    vtab = gtNewOperNode(GT_IND, TYP_I_IMPL, thisPtr);
+    vtab->gtFlags |= GTF_IND_NONFAULTING;
+    vtab->gtFlags |= GTF_IND_INVARIANT;
+
+    // Get the appropriate vtable chunk
+    if (vtabOffsOfIndirection != CORINFO_VIRTUALCALL_NO_CHUNK)
+    {
+        if (isRelative)
+        {
+            // MethodTable offset is a relative pointer.
+            //
+            // Additional temporary variable is used to store virtual table pointer.
+            // Address of method is obtained by the next computations:
+            //
+            // Save relative offset to tmp (vtab is virtual table pointer, vtabOffsOfIndirection is offset of
+            // vtable-1st-level-indirection):
+            // tmp = vtab
+            //
+            // Save address of method to result (vtabOffsAfterIndirection is offset of vtable-2nd-level-indirection):
+            // result = [tmp + vtabOffsOfIndirection + vtabOffsAfterIndirection + [tmp + vtabOffsOfIndirection]]
+            //
+            //
+            // If relative pointers are also in second level indirection, additional temporary is used:
+            // var1 = vtab
+            // var2 = var1 + vtabOffsOfIndirection + vtabOffsAfterIndirection + [var1 + vtabOffsOfIndirection]
+            // result = [var2] + var2
+            //
+            unsigned varNum1 = lvaGrabTemp(true DEBUGARG("var1 - vtab"));
+            unsigned varNum2 = lvaGrabTemp(true DEBUGARG("var2 - relative"));
+            GenTree* asgVar1 = gtNewTempAssign(varNum1, vtab); // var1 = vtab
+
+            // [tmp + vtabOffsOfIndirection]
+            GenTree* tmpTree1 = gtNewOperNode(GT_ADD, TYP_I_IMPL, gtNewLclvNode(varNum1, TYP_I_IMPL),
+                                              gtNewIconNode(vtabOffsOfIndirection, TYP_INT));
+            tmpTree1 = gtNewOperNode(GT_IND, TYP_I_IMPL, tmpTree1, false);
+            tmpTree1->gtFlags |= GTF_IND_NONFAULTING;
+            tmpTree1->gtFlags |= GTF_IND_INVARIANT;
+
+            // var1 + vtabOffsOfIndirection + vtabOffsAfterIndirection
+            GenTree* tmpTree2 = gtNewOperNode(GT_ADD, TYP_I_IMPL, gtNewLclvNode(varNum1, TYP_I_IMPL),
+                                              gtNewIconNode(vtabOffsOfIndirection + vtabOffsAfterIndirection, TYP_INT));
+
+            // var1 + vtabOffsOfIndirection + vtabOffsAfterIndirection + [var1 + vtabOffsOfIndirection]
+            tmpTree2         = gtNewOperNode(GT_ADD, TYP_I_IMPL, tmpTree2, tmpTree1);
+            GenTree* asgVar2 = gtNewTempAssign(varNum2, tmpTree2); // var2 = <expression>
+
+            // ToDo: Is this indirection Invariant or not?
+            // -- probably not since it is the last indirection on this path
```

Comment:
Right, for coreclr the method pointer fetches should not be invariant, but can be nonfaulting.

So, this should be nonfaulting.